include/internal/quic_record_util.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_QUIC_RECORD_UTIL_H
+# define OSSL_QUIC_RECORD_UTIL_H
+
+# include <openssl/ssl.h>
+
+/*
+ * QUIC Key Derivation Utilities
+ * =============================
+ */
+
+/* HKDF-Extract(salt, IKM) (RFC 5869) */
+int ossl_quic_hkdf_extract(OSSL_LIB_CTX *libctx,
+                           const char *propq,
+                           const EVP_MD *md,
+                           const unsigned char *salt, size_t salt_len,
+                           const unsigned char *ikm, size_t ikm_len,
+                           unsigned char *out, size_t out_len);
+
+/* HKDF-Expand(PRK, info, L) (RFC 5869) */
+int ossl_quic_hkdf_expand(OSSL_LIB_CTX *libctx,
+                          const char *propq,
+                          const EVP_MD *md,
+                          const unsigned char *secret,
+                          const unsigned char *info, size_t info_len,
+                          unsigned char *out, size_t out_len);
+
+/* HKDF-Expand-Label(Secret, Label, Context, Length) (RFC 8446 - TLS 1.3) */
+int ossl_quic_hkdf_expand_label(OSSL_LIB_CTX *libctx,
```

Comment:
Almost &mdash; the problem is that `tls13_hkdf_expand` depends on a `SSL_CONNECTION` object.

Maybe it could be refactored? The only other concern I have is use of `ERR_raise`. Since this function can be triggered from unauthenticated network traffic. (Basically, since in QUIC, key update is triggered by the key phase bit in the header, an attacker can initiate key update attempts by flipping the key phase bit randomly. This will not ultimately succeed but we do have to do some key derivation first to determine whether the key update is legitimate or not.) Although those are only internal/allocation errors, so I guess nothing to really be concerned about.

`tls13_hkdf_expand` could be refactored to call another function with `sctx->libctx` and `sctx->propq`, allowing QRL to also use it? Works for me.

Reply:
Refactoring it makes sense to me. Having 2 implementations of essentially the same thing seems like a bad idea to me. We should reuse where possible/practical.